tensorboard/util/timing.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for measuring elapsed time."""
+
+import contextlib
+import logging
+import threading
+import time
+
+from tensorboard.util import tb_logging
+
+logger = tb_logging.get_logger()
+
+
+def log_latency(region_name_or_function_to_decorate, log_level=None):
+    """Log latency in a function or region.
+
+    Three usages are supported. As a decorator:
+
+    >>> @log_latency
+    ... def function_1():
+    ...     pass
+    ...
+
+
+    As a decorator with a custom label for the region:
+
+    >>> @log_latency("custom_label")
+    ... def function_2():
+    ...     pass
+    ...
+
+    As a context manager:
+
+    >>> def function_3():
+    ...     with log_latency("region_within_function"):
+    ...         pass
+    ...
+
+    Args:
+        region_name_or_function_to_decorate: Either: a `str`, in which
+            case the result of this function may be used as either a
+            decorator or a context manager; or a callable, in which case
+            the result of this function is a decorated version of that
+            callable.
+        log_level: Optional integer logging level constant. Defaults to
+            `logging.INFO`.
+
+    Returns:
+        A decorated version of the input callable, or a dual
+        decorator/context manager with the input region name.
+    """
+
+    if log_level is None:
+        log_level = logging.INFO
+
+    if isinstance(region_name_or_function_to_decorate, str):
+        region_name = region_name_or_function_to_decorate
+        return _log_latency(region_name, log_level)
+    else:
+        function_to_decorate = region_name_or_function_to_decorate
+        qualname = getattr(function_to_decorate, "__qualname__", None)
+        if qualname is None:
+            qualname = str(function_to_decorate)
```

Comment:
It’s not super easy to get a handle to a function without `qualname` to
test this case, but with at least current versions of numpy the
following suffices as a manual test:

```python
>>> import logging
>>> import numpy as np
>>> from tensorboard.util import timing
>>>
>>> np.__version__
'1.18.2'
>>> np.sin.__qualname__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'numpy.ufunc' object has no attribute '__qualname__'
>>> sin = timing.log_latency(np.sin, log_level=logging.WARNING)
>>> sin([1, 2, 3])
MainThread[7f81fc8d7740] ENTER <ufunc 'sin'>
MainThread[7f81fc8d7740] LEAVE <ufunc 'sin'> - 0.000302s elapsed
array([0.84147098, 0.90929743, 0.14112001])
```

(Don’t want to commit this in case [the upstream bug][1] is ever fixed.)

[1]: https://github.com/numpy/numpy/issues/4952